### PR TITLE
Add boss hiscores

### LINKED ISSUE #10397: Add support for Boss Hiscores
With the update on the 5th of December, Jagex implemented Boss Hiscores. These should be incorporated into the Hiscore panel.

Exactly how it should look will need some discussion.
Currently, there is basically no space in the current UI to display both the Hiscores and the Boss Hiscores at the same time, since the normal Hiscores fill the entire ~503 pixel height when the client is its smallest.
Having the panel scroll would be a simple fix, but since the panel hasn't ever scrolled before, it could confused users. A button also poses a similar problem, since nothing else is a button on there.

The layout of `index_lite` should be pretty simple to modify, I've mapped out the new format [here](https://pastebin.com/xrN6ZnUv). Note: Callisto and Cerberus are missing from the list. I have no idea when Jagex will notice and fix it.

### PR DESCRIPTION
**Pull request by Bram91**
Closes #10397 
Add support for boss kc in the hiscore panel also some layout tweaks to make everything align properly and to format numbers over a 1000(for boss kc,clue scroll completion and league points).
hiscores

When the client is at the minimum size it will have to scroll, currently there is no clear indication for that besides partial boss icons at the bottom of the panel and the appearance of a scroll bar to the side:
Capture

Suggestions for the latter are more than welcome

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java [hunk 1]
```diff
@@ -284,6 +380,38 @@ public void mousePressed(MouseEvent mouseEvent)
 
 		add(minigamePanel, c);
 		c.gridy++;
+
+		bossPanel = new JPanel();
+		bossPanel.setLayout(new GridLayout(15, 3));
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java]:
```suggestion
		bossPanel.setLayout(new GridLayout(0, 3));
```

File: runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java [hunk 2]
```diff
@@ -296,9 +424,11 @@ public void onActivate()
 	/* Builds a JPanel displaying an icon and level/number associated with it */
 	private JPanel makeSkillPanel(HiscoreSkill skill)
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java]: This should get a better name

File: runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java [hunk 3]
```diff
@@ -307,28 +437,62 @@ private JPanel makeSkillPanel(HiscoreSkill skill)
...
+			skillPanel.setBorder(new EmptyBorder(7, 15, 7, 15));
+		}
+
+		skillPanel.add(labelIcon);
+		skillPanel.add(labelText);
 
 		return skillPanel;
 	}
 
+	private boolean checkBoss(HiscoreSkill skill)
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java]:
```suggestion
	private boolean isBoss(HiscoreSkill skill)
```

File: runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java [hunk 4]
```diff
@@ -307,28 +437,62 @@ private JPanel makeSkillPanel(HiscoreSkill skill)
...
+	{
+		if (skill != null)
+		{
+			for (HiscoreSkill temp : BOSSES)
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java]: this should just be a lookup in a set or better yet a field on the enum

File: runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java [hunk 5]
```diff
@@ -438,11 +602,37 @@ else if (!isSkill || exp != -1L)
 
 				if (level != -1)
 				{
-					label.setText(Integer.toString(level));
+					// Format numbers to prevent overflowing the labels
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java]: Use one of the standard text formatters for this

File: runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java [hunk 6]
```diff
@@ -466,6 +656,10 @@ private HiscoreSkill find(int index)
 		{
 			return SKILLS.get(index);
 		}
+		else if (index >= (SKILLS.size() + 7))
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java]: This is awful

File: runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java [hunk 7]
```diff
@@ -64,35 +66,78 @@
 import net.runelite.http.api.hiscore.HiscoreEndpoint;
 import net.runelite.http.api.hiscore.HiscoreResult;
 import net.runelite.http.api.hiscore.HiscoreSkill;
+import static net.runelite.http.api.hiscore.HiscoreSkill.ABYSSAL_SIRE;
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java]: this should be a * import at this point

File: runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java (hunk 2 above)
Comment (Bram91) [reply runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java]: renamed it to makeHiscorePanel

File: runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java (hunk 5 above)
Comment (Bram91) [reply runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java]: The available formatters dont allow the type of formatting required, in some situations the numbers would be displayed partially

**Comment by Bram91** [comment]
Removed the button and made the requested changes

File: runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java [hunk 8]
```diff
@@ -307,28 +323,56 @@ private JPanel makeSkillPanel(HiscoreSkill skill)
...
 		JPanel skillPanel = new JPanel();
 		skillPanel.setBackground(ColorScheme.DARKER_GRAY_COLOR);
-		skillPanel.setBorder(new EmptyBorder(2, 0, 2, 0));
-		skillLabels.add(label);
-		skillPanel.add(skillLabels.get(skillLabels.size() - 1));
+
+		GridLayout grid = new GridLayout(1, 2);
+		skillPanel.setLayout(grid);
+		//Set different spacings based on skill or bosses
+		if (BOSSES.contains(skill))
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java]:
```suggestion
		if (skill != null && skill.getType() == SkillActivityType.BOSS)
```

File: runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java [hunk 9]
```diff
@@ -307,28 +323,56 @@ private JPanel makeSkillPanel(HiscoreSkill skill)
...
+		labelText.setIconTextGap(totalLabel ? 10 : 4);
 		JPanel skillPanel = new JPanel();
 		skillPanel.setBackground(ColorScheme.DARKER_GRAY_COLOR);
-		skillPanel.setBorder(new EmptyBorder(2, 0, 2, 0));
-		skillLabels.add(label);
-		skillPanel.add(skillLabels.get(skillLabels.size() - 1));
+
+		GridLayout grid = new GridLayout(1, 2);
+		skillPanel.setLayout(grid);
+		//Set different spacings based on skill or bosses
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java]: why?

File: runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java [hunk 10]
```diff
@@ -438,11 +482,37 @@ else if (!isSkill || exp != -1L)
...
 				if (level != -1)
 				{
-					label.setText(Integer.toString(level));
+					// Format numbers to prevent overflowing the labels
+					if (skill != OVERALL)
+					{
+						String levelText;
+						if (level >= 1000)
+						{
+							Long divideBy = 1_000L;
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java]:
```suggestion
							long divideBy = 1_000L;
```

File: runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java [hunk 11]
```diff
@@ -485,7 +555,7 @@ private HiscoreSkill find(int index)
 				return BOUNTY_HUNTER_HUNTER;
 		}
 
-		return null;
+		return BOSSES.get(index - (SKILLS.size() + 7));
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java]: this is still awful; I have no idea what this SKILLS+7 is for

File: runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java [hunk 12]
```diff
@@ -584,27 +658,43 @@ private String detailsHtml(HiscoreResult result, HiscoreSkill skill)
...
 				default:
 				{
-					Skill requestedSkill = result.getSkill(skill);
-					final long experience = requestedSkill.getExperience();
-
-					String rank = (requestedSkill.getRank() == -1) ? "Unranked" : QuantityFormatter.formatNumber(requestedSkill.getRank());
-					String exp = (experience == -1L) ? "Unranked" : QuantityFormatter.formatNumber(experience);
-					String remainingXp;
-					if (experience == -1L)
+					if (BOSSES.contains(skill))
```
Comment (abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java]:
```suggestion
					if (skill.getType() == SkillActivityType.BOSS)
```

File: runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java [hunk 13]
```diff
@@ -170,7 +170,7 @@ public HiscorePanel(HiscoreConfig config)
 		searchBar.setPreferredSize(new Dimension(PluginPanel.PANEL_WIDTH - 20, 30));
 		searchBar.setBackground(ColorScheme.DARKER_GRAY_COLOR);
 		searchBar.setHoverBackgroundColor(ColorScheme.DARK_GRAY_HOVER_COLOR);
-		searchBar.setMinimumSize(new Dimension(0, 30));
+		searchBar.setMinimumSize(new Dimension(188, 30));
```
[the remark this answers is not in the document]
Comment (Bram91) [reply runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java]: Missed that line when reverting the button change

File: runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java (hunk 9 above)
Comment (Bram91) [reply runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java]: To make the icons align properly regardless of the kc number when having bosses at 1kc vs bosses with 100k kc the layout got very wonky

File: runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java (hunk 11 above)
Comment (Bram91) [reply runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java]: the skillLabels for boss kcs start at position `SKILLS.size() + 7`  in the skillLabels List so I need to lower the index by that amount to retrieve the respective boss in the BOSSES List

**Pull request merged**